librz/util/float/float.c

```diff
@@ -2197,6 +2243,296 @@ RZ_API RZ_OWN RzFloat *rz_float_trunc(RZ_NONNULL RzFloat *f) {
+	rz_return_val_if_fail(f, NULL);
+
+	RzBitVector *ret = rz_bv_new(length);
+	RzBitVector *tmp, *rounded;
+	ut32 exp = float_exponent(f);
+	RzFloatFormat format = f->r;
+	bool sign = get_sign(f->s, format);
+	ut32 bias = rz_float_get_format_info(format, RZ_FLOAT_INFO_BIAS);
+	bool is_subnormal = exp == 0;
+	st32 exp_no_bias = is_subnormal ? (1 - bias) : (exp - bias);
+	ut32 total_len = rz_float_get_format_info(format, RZ_FLOAT_INFO_TOTAL_LEN);
+	ut32 man_len = rz_float_get_format_info(format, RZ_FLOAT_INFO_MAN_LEN);
+
+	// rounding float to get an integer means
+	// we should try to reserve `exponent` bits of mantissa
+	// drop extra bits or append zeros
+	// 1.MM..M * 2^exp = 1MM..M * 2^0 (integer)
+	bool should_inc = false;
+	RzBitVector *sig = rz_float_get_mantissa(f);
+
+	// sub normal one has no hidden bit, others should set to 1
+	if (!is_subnormal) {
+		rz_bv_set(sig, man_len, true);
+	}
+
+	if (exp_no_bias >= 0) {
+		// has `exp_no_bias` + 3 + 1 length
+		tmp = round_significant(sign, sig, exp_no_bias, mode, &should_inc);
+	} else {
+		// float 1.M..M * 2^exp, when exp < 0
+		// flatten it and we have 0.0..1M..M (|exp|+1 zeros before 1MMM...)
+		// set a fake 1 before radix point, and we can use round_significant to round
+		ut32 remained_zeros = total_len - man_len - 1;
+		RzBitVector *fake_f;
+		if (-exp_no_bias > remained_zeros) {
+			// prepend
+			fake_f = rz_bv_prepend_zero(sig, -exp_no_bias - remained_zeros);
+		} else {
+			fake_f = rz_bv_dup(sig);
+		}
+		rz_bv_set(fake_f, rz_bv_len(fake_f) - 1, true);
+		tmp = round_significant(sign, fake_f, 0, mode, &should_inc);
+
+		// unset the fake 1 in tmp
+		// tmp has 3 + 1 + precision = 4
+		rz_bv_set(tmp, 0, false);
+		rz_bv_free(fake_f);
+	}
+	rz_bv_free(sig);
+	sig = NULL;
+
+	// rounded result
+	if (should_inc) {
+		// WARN: possible overflow => no enough length
+		RzBitVector *bv_one;
+		bv_one = rz_bv_new_one(rz_bv_len(tmp));
+		rounded = rz_bv_add(tmp, bv_one, NULL);
+		rz_bv_free(bv_one);
+	} else {
+		rounded = rz_bv_dup(tmp);
+	}
+	rz_bv_free(tmp);
+	tmp = NULL;
+
+	// assume we r handling absolute value
+	// now for negative, convert it to 2's complement
+	if (sign) {
+		// to keep it an negative, make ret all set to bit 1
+		rz_bv_toggle_all(ret);
+		tmp = rz_bv_complement_2(rounded);
+		rz_bv_free(rounded);
+		rounded = tmp;
+		tmp = NULL;
+	}
+
+	// WARN: possible overflow if length < exp_no_bias
+	// WARN: higher bits may be cut off
+	rz_bv_copy_nbits(rounded, 0, ret, 0, rz_bv_len(rounded));
+	rz_bv_free(rounded);
+	return ret;
+}
+
+/**
+ * convert float from format A to a new format B
+ * \param f float
+ * \param format new format
+ * \param mode rounding mode
+ * \return converted float with format B
+ */
+RZ_API RZ_OWN RzFloat *rz_float_convert(RZ_NONNULL RzFloat *f, RzFloatFormat format, RzFloatRMode mode) {
+	rz_return_val_if_fail(f, NULL);
+
+	// TODO: preprocess special float here
```

Comment:
What do you mean by "special float"?

Reply:
Nan, Inf ... converts them into different format